[FIX] http_routing, website: prevent crash when using `fw` in url

rdeodoo opened the pull request:
Before this commit, the routing map generated and used would be the one from
the website the request is performed, instead of the one from the `fw` website
ID which will be the one we redirect the user.

This issue was introduced with the routing map by website, be8fc2296b38 and is
restricted to a single case: a publisher using the website switcher, and it
won't happen on next page naviguation/refresh as the `fw` website id will be
the same as the current website's ID.

---------------------------------------------
**Technical analysis:**

_website/models/ir_http.py_
```python
_dispatch
  request.website_routing = request.env['website'].get_current_website().id
  response = super(Http, cls)._dispatch()
  return response

_add_dispatch_parameters
  if mw + publisher + fw in url:
    request.env['website']._force_website(request.httprequest.args.get('fw'))
  super(Http, cls)._add_dispatch_parameters(func)

_match
  key = key or (request and request.website_routing)
  return super(Http, cls)._match(path_info, key=key)

_generate_routing_rules
  website_id = request.website_routing
  logger.debug("_generate_routing_rules for website: %s", website_id)
```
_http_routing/models/ir_http.py_
```python
_dispatch
  rule, arguments = cls._match(request.httprequest.path)
  cls._authenticate(func.routing['auth'])
  cls._add_dispatch_parameters(func)
```
_base/models/ir_http.py_
```python
_match
  return cls.routing_map(key=key).bind('').match(
    return_rule=True,
    method=request.httprequest.method,
    path_info=path_info
  )

routing_map
  for url, endpoint, routing in cls._generate_routing_rules(mods, converters=cls._get_converters()):
```

Current flow when accessing /?fw=1 depuis le website 2
1. website_routing is set to get_current_website -> 2 in _dispatch
2. _match is called in _dispatch
3. _match is calling routing_map with key = website_routing (set to 2 in step 1.)
4. routing_map is calling _generate_routing_rules which generate it based on website_routing (set to 2 in step 1.)

5. user is retrieved/authenticated in _dispatch by calling _authenticate
6. in _dispatch, _add_dispatch_parameter is called, where fw in url is forced in session and so get_current_website now return correct website_id -> 1

**Fix**: move step 6 before step 1 -> force fw in session before setting website_routing to the wrong website id
**Problem**: we need the user to be authenticated to use env.user.has_group(designer + mw) to allow fw to be forced
**Suggested** solution: force fw before user is authenticated and store the fact we did it, then when authenticating, if not the rights, redirect to previous website

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/69416).

rdeodoo commented:
@JKE-be I didn't bother moving the code in website where it should be, before we validate the change.
This is just my analysis & suggested solution to be discussed :+1:

rdeodoo commented on addons/http_routing/models/ir_http.py:
```diff
@@ -459,6 +463,7 @@ def _dispatch(cls):
                 cls._auth_method_public()
         except Exception as e:
             return cls._handle_exception(e)
+        cls._dispatch_post_auth_hook()
```
Can't think of a better way than this hook
I'd have prefer to do that in `_add_dispatch_parameter` but this is called only on is_frontend requests :/ that would also have allow to directly call request.redirect

JKE-be replied:
in which case we would like to support fw for a non is_frontend route ? 
Reminder, this option is only useful during dev. Not really in prod... 
In worst case, it is acceptable for me to switch to / instead of current page if you need to be sure that the switch is done on a is_frontend route.

JKE-be replied:
As discuss in pm, let try to move it in external controller. ty

rdeodoo replied:
> in which case we would like to support fw for a non is_frontend route ? [..] if you need to be sure that the switch is done on a is_frontend route.

You didn't understand, that's the opposite, I'd like that behavior to only happen on frontend route, but in the current PR state, we don't know yet if we are in frontend route at that point in the dispatch.
So can't put the second part of the code in add_dispatch_parameter as only called in frontend route, so I created a hook that is called for every route

But I will push in a few minutes a diff to use a route instead as discussed, that's a bigger diff that you imagine IMO, there is a branch for design-theme, and if we drop the `fw` support in 13.0, also in upgrade repo (for tests).

If we want to keep the retrocompatibility in 13.0 for FW, the diff will still have to be applied in master even in design theme, but I don't think we should keep retro here

rdeodoo reviewed (commented):
Added a test also

rdeodoo commented:
I will check runbot tomorrow, maybe something is missing but here is a summary:
1. I dropped the existing force website route, we don't use it since https://github.com/odoo-dev/odoo/commit/569554b9e035c638d57b63740cd9141634508a14
2. I dropped the fw parameter behavior
3. We need to adapt design-theme tests -> https://github.com/odoo-dev/design-themes/commit/54882ecab7f49c070430d87c7a548f1b180c7988
4. We need to adapt upgrade tests if we drop the `fw` in 13.0 -> https://github.com/odoo/upgrade/blame/master-imp-favorite-product-rde/migrations/test_themes/tests/test_themes_loading.py#L48 I have the diff in local, I'll push if you confirm ok to drop fw

rdeodoo commented:
Also, for an easier history comprehension in case someone blame and arrive on this PR, here is the `fw` history so far:
1. Introduce with https://github.com/odoo-dev/odoo/commit/4f6ec1cd2a9253bb0c6c17d304490c7809facf8a#diff-fe005825c9360edbc616517111aa67926161221c42003cad29f88a0214cd909eR289
2. Fixed here to work with multi domain https://github.com/odoo-dev/odoo/commit/f62003952784a41f3dcb7bf8a6a0c3a23296a047
3. Fixed here to work with domain + no domain https://github.com/odoo-dev/odoo/commit/569554b9e035c638d57b63740cd9141634508a14
4. Fixed public user being logged in from the previous website https://github.com/odoo-dev/odoo/commit/9c4eab3ef475b5d04d4944e335cb7f7b41235995
5. Fixed to work with scheme & ports https://github.com/odoo-dev/odoo/commit/964aa3d6965bc1cc8757e64cefd9535e991aa1df

rdeodoo commented:
TODO in master remove `fw` support & don't forget to remove `fw` in theme

JKE-be commented:
ty

@robodoo r+

JKE-be reviewed (approved):
lgtm
